### DISCUSSION TITLE
Add the ability for metadata sources to use an artists "main/canonical" name instead of the name that's credited on the work

### DISCUSSION THREAD

**Discussion by Alariius** [beetbox/beets#5500] [Ideas]
On Musicbrainz and Discogs (and possibly others) there is a lot of flexibility in how an artists name can be credited on a specific work. The name can be capitalized differently, use an alias or multiple artists can use weird and non-standard joining characters between their names. 
This can result in either a single artist appearing multiple times in a library under different names or multiple artists (or multiple spellings of an artists name) being interpreted as a single artist. 

I'm currently dealing with this by changing the artist name manually after importing in cases like this. It would be great however if there was the option for a metadata source to look up an artists main name and use that one instead.

Since the aliases and different spellings of an artists name still use the same ID and link to the same entry on Musicbrainz and Discogs I don't imagine this would be too difficult to implement.

If there already is such a feature/plugin or a way to achieve similar results I would appreciate if someone could tell me how.

**Comment by RollingStar** [comment]
Related: https://discourse.beets.io/t/plugin-idea-expose-also-performs-as-for-album-artist-field/1541

**Comment by bossanova808** [comment]
Thoroughly concur with this suggestion - very desirable feature.  (I.e. as per Picard's 'use standardizes artist's names' option).

Practical cases - I want e.g. all Prince under Prince (both in my library and on disk), not as:

Prince
Prince & The New Power Generation
Prince and The New Power Generation
Prince and The Revolution
 
..as I am currently getting.

**Reply by RollingStar** [reply]
Can you clarify the Prince situation? Would most/all Prince fans consider that all just "Prince"? For bonus credit, find the API info like this.

https://musicbrainz.org/ws/2/artist/ace273a8-a660-4bce-8d8f-569502346e4b?inc=artist-rels&fmt=json
https://musicbrainz.org/ws/2/artist/75ade3f1-275f-4f8e-b26a-48d391127dd5?inc=artist-rels&fmt=json

> per Picard's 'use standardizes artist's names' option

There's a starting point. Picard pulls from MusicBrainz, so where is it getting that?

**Reply by bossanova808** [reply]
I'd say yes, by and large people regard it as Prince + his band of the time.  
If you're in a record shop, you're looking for all of it under Prince.  
But there's also no doubt that e.g. Purple Rain is by Prince and the Revolution...so honestly, l can also see this getting really difficult/ambiguous quickly, and very debatable.  

So perhaps the current re-write (or substitute) plugins are the most sensible options?

As far as I can see there are a few current options for managing this:
* `beet modify` the values to how I would like then - but this will then be lost if I e.g. run a `beet mbsync` on my library (there's no way to  'protect' modified values is there?)
* the `substitute` plugin - but that only lets me clean up the paths side of things
* the `rewrite` plugin - which is what I am using, and puts things under the one folder, but in my library the artist's are still separate (which 

Musicbrainz _lists_ all the works under Prince here:
https://musicbrainz.org/artist/070d193a-845c-479f-980e-bef15710653e

...and, given that, that's really what I'd like to see in my collection (both on disk, but particularly when browsing my library in Lyrion Music Server AKA LMS) - so if I want to find Purple Rain, I just click on Prince from my artists list...rather than have to remember which band did what.

But the releases are credited to the various bands (and often various spellings of various bands, sigh).

In Picard, it is (now) default to use Standardized Artist Names (May 2023):
https://github.com/metabrainz/picard/blob/154aab5cbb65f1f02783c36d57ef7038b9689cd5/NEWS.md?plain=1#L221

And here are maybe some key bits of Picard code:
https://github.com/metabrainz/picard/blob/154aab5cbb65f1f02783c36d57ef7038b9689cd5/picard/mbjson.py#L249
https://github.com/metabrainz/picard/blob/154aab5cbb65f1f02783c36d57ef7038b9689cd5/picard/mbjson.py#L358

From the API at  quick glance, it's just the `name` - but of the artist, not the release I guess?
https://musicbrainz.org/ws/2/artist/070d193a-845c-479f-980e-bef15710653e?inc=artist-rels&fmt=json

(I thought Picard was tagging all these as Prince, but have moved over to beets entirely now, and with a quck test if I tag these with Picard, it is still Prince and The Revolution (etc) it seems, so I'm not even right about the Picard behaviour here, actually).

It's soon a can of worms but I do think the behaviour is desirable...I'd happily modify things after the initial import if I could still run things like `mbsync` and not lose all the work...but that doesn't seem possible.

**Reply by RollingStar** [reply]
Tons of good comments here

https://tickets.metabrainz.org/browse/PICARD-2673

https://community.metabrainz.org/t/artist-aliases-for-past-names-vs-picards-use-standardized-artist-names-option/642755/13

https://tickets.metabrainz.org/browse/PICARD-2634

> However, a clean solution would be to actually change the option itself, and not just its default setting. Instead of a simple boolean, there should be three possible settings:
> 1. Never standardize artist names. This will use the artist credit names as before.
> 2. Standardize artist names except in case of an artist name change. This should be the default setting. It will use the standardized artist names by default. However, if there exists a different primary artist alias which was active at the time when the release appeared, then that artist alias is used instead.
> 3. Always standardize artist names. This will always use the standardized artist names as before.

There are questions in the comments on whether MB has thorough enough data to make # 2 viable.

**Reply by bossanova808** [reply]
Yeah I can see this is very complicated in practice.

I think for me, managing these manually through `rewrite` is probably ultimately the best choice - certainly this means my folders get consolidated - although contrary to the docs, I am not seeing the metadata/db change with this, so my library views are still frustratingly not how I want them.

And am also seeing beets do what seems to be a spurious disambiguation in this case, as these re-written albums have e.g. `[album]` appended to the folder name, for reasons I can't understand.  

So it's _sort of_ working with beets and `rewrite`, but far from ideal... 

Honestly in the end, for this finite number of things, it might just be better if I move these special editions to my 'non Musicbrainz' area for troublesome things and simply manually so these ones with mp3tag.  I might just have to admit that any massive sorting project comes with some compromises...

**Reply by snejus** [reply]
@bossanova808 how does your `rewrite` configuration look like?

Have you tried using the [`inline`](https://beets.readthedocs.io/en/latest/plugins/inline.html) plugin? It may provide a bit more flexibility for your use case. See how I use it to resolve paths:

```yaml
plugins:
  - inline
...
item_fields:
  has_lyrics: bool(lyrics)
  label_or_albumartist: |
    aartist = (albumartist or artist).split(", ")[0]
    if not label or label.lower() in {"not on label", aartist.lower()}:
        return f"self_{aartist}"
    return label or albumartist or artist
  singleton_track_artist: '"" if label.lower() == artist.lower() else f"{artist}_"'
  track_artist: '"" if label.lower() == artist.lower() or albumartist == artist else f"{artist}_"'
  album_name: |
    return "_".join((a for a in [
      catalognum if catalognum != album else "",
      (albumartist if albumartist not in {label, "Various Artists"} else ""),
      album
    ] if a.strip()))
  track_identification: |
    if (track_alt := globals().get("track_alt")) and track_alt[0].isalpha():
      idx = track_alt
    elif track:
      idx = str(track).zfill(2)
    else:
      return ""

    if disctotal > 1:
      idx = f"{disc}_{idx}"

    return f"{idx}_"
  withdrawn: albumstatus == "Withdrawn"
paths:
  albumtype:broadcast: |
    sets/
    %lower{
      %if{
        $album,
        $album/$track_identification$track_artist$title,
        $track_identification$track_artist$title
      }
    }
  singleton: |
    %lower{
      $label_or_albumartist/
      %if{$withdrawn,_withdrawn/}
      %if{$catalognum,${catalognum}_}
      $singleton_track_artist$title
    }
  comp: |
    %lower{
      $label_or_albumartist/
      %if{$withdrawn,_withdrawn/}
      $album_name-%aunique{}/
      $track_identification${artist}_$title
    }
  default: |
    %lower{
      $label_or_albumartist/
      %if{$withdrawn,_withdrawn/}
      $album_name-%aunique{}/
      $track_identification${track_artist}$title
    }
```

**Reply by bossanova808** [reply]
I have not ventured towards `inline` yet - looks powerful, but also some work and fiddly.  Would need to create a test library etc.  But thanks, will keep it in mind.

`rewrite` looks like a series of basic rewrites to consolidate various things:

```
rewrite:
    artist Beth Gibbons \& Rustin Man: Beth Gibbons
    artist Charles Mingus Sextet with Eric Dolphy: Charles Mingus
    artist David Bowie \/ Trevor Jones: David Bowie
    artist Duke Ellington and His Orchestra: Duke Ellington
    artist Explosions in the Sky \& David Wingo: Explosions in the Sky
    artist Flock of Dimes \/ Madeline Kenney: Flock of Dimes
    artist Hans Zimmer \& Benjamin Wallfisch: Hans Zimmer
    artist Iron \& Wine \& Ben Bridwell: Iron \& Wine
    artist The Jimi Hendrix Experience: Jimi Hendrix
    artist John Coltrane Quartet: John Coltrane
    artist Johnny Flynn \& Robert Macfarlane: Johnny Flynn
    artist José González \& The String Theory: José González
    artist Julie Byrne & Laugh Cry Laugh: Julie Byrne
    artist Lord Huron \& Ben Schneider: Lord Huron
    artist Moondog Jr\.: Zita Swoon
    artist Neko Case \& Her Boyfriends: Neko Case
    artist PJ Harvey 1& John Parish: PJ Harvey
    artist Prince \& The New Power Generation: Prince
    artist Prince and The New Power Generation: Prince
    artist Prince and The Revolution: Prince
    artist Ray LaMontagne \& The Pariah Dogs: Ray LaMontagne
    artist Sting \/ The Police: Sting
    artist Vince Jones \/ Grace Knight: Vince Jones
    ```

**Reply by snejus** [reply]
This looks very neat! If I was to generalize this using `inline`, I'd probably split artists on **` & `** and **` / `** and use the first one. Though this would apply to all artists in the database, not only those in the list.

**Reply by bossanova808** [reply]
Thanks, it is working pretty well - I think a generalised split on `&/` would be a bridge too far, with too many unintended consequences unfortunately.

While I have you, what I read around here suggests that the metadata should also be updated by `rewrite` operations (i.e. I'd expect `albumartist` to change to e.g. Prince).  But...that's not the case.  So it's really only handling the folder side of things.

Looking at the history it seems the `rewrite` changing metadata was actually originally unintended...has that been 'fixed'?  In which case, one would have to `modify` such data - but then it will be over-written with e.g. an `mbsync` right?   

Can you think of any resilient way to achieve this sort of thing?
